Undo v1.25.2

undo-bkenward opened the pull request:
For review.

TestIssue4051 in the `basic` test set fails with both `rr` and `undo` backends, with:
```
expected error "package main has no function Hello", got "call from within the Go runtime"
```

dpapastamos committed:
Support for Undo record/replay backend.

* Update the documentation to mention UDB or LiveRecorder alongside rr.
* Add makefile support for testing the Undo backend.
* Add DLV_RECORD_REPLAY_BACKEND environment variable which overrides
  the record-replay backend, for use in GoLand.
* Implement checkpoints as a map from checkpoint number to time.
* Suppress resume call after restart.
* Implement "when" command using "get_time" serial command.
* Implement call injection using "set_debuggee_volatile".
* Implement "restart" using "goto_time".
* Enable or disable Undo backend appropriately in test cases.

dpapastamos committed:
Remove build status information from README.

gareth-rees committed:
Improve coverage of record-replay backends by allowing recording.

gareth-rees committed:
Rename import paths to point to our fork.

gareth-rees committed:
Improved method for reading the G pointer on x86.

In Undo, debuggee state is immutable during replay, so reading the G
pointer via code injection won't work; and in any case, this approach
is simpler.

gareth-rees committed:
Exclude TestCallFunction cases that fail with the Undo backend.

mark-undoio committed:
Implement basic progress indicator support in Delve.

The error handling here was a bit painful but only needs doing in these two
places (for now).

mark-undoio committed:
Implement save / load of checkpoints with Undo v7.0+ session files.

This is made compatible with UDB's bookmarks by using the UDB bookmark name as
the "Note" field in Delve. To save Delve checkpoints, the "Note" must be made
unique, so as to be a valid bookmark name in UDB.

undo-bkenward committed:
pkg/proc/gdbserial: allow extra detail in ErrBackendUnavailable

mark-undoio committed:
Add a --symbol-file flag to "delve replay" for external symbols.

This allows a separate debug symbol file to be specified when replaying a
recording with the Undo backend.

This can be used to supply debug information when the main binary has been
stripped.

mark-undoio committed:
A script to install Undo's dlv into a GoLand installation.

mark-undoio committed:
Update serial protocol get_recording_ids support for recent master.

Since we added support for a process tree level UUID the
`vUDB;get_recording_ids` serial protocol message can return 4 UUIDs.

It's easiest to just handle this unconditionally for now, we can clean up the
legacy support later.

mark-undoio commented on service/rpc1/client.go:
```diff
@@ -0,0 +1,320 @@
+package rpc1
```
Are we accidentally reviving their old RPC protocol here in the process of bringing our patches across?  I thought these had disappeared in the previous rebase but maybe I accidentally reanimated something.

undo-bkenward replied:
Good spot. I'll purge that and squash it into the "rename import path" commit that's currently re-adding these.

mark-undoio commented on service/test/integration1_test.go:
```diff
@@ -1,1096 +0,0 @@
-package service_test
```
Test of the old RPC protocol disappears here agin.

mark-undoio commented on pkg/proc/gdbserial/undo.go:
```diff
@@ -192,8 +192,10 @@ func getSessionPath(conn *gdbConn) (string, error) {
 	uuids := strings.Split(recording_ids, ";")
-	if len(uuids) != 3 || uuids[1] == "" {
-		panic("unexpected response from get_recording_ids")
+	// We may get 3 or 4 UUIDs depending on the version of the server, since more recent
```
I think you could reasonably just squash this into the commit that adds `undo.go` at this point.

I usually do a quick squash down of these kinds of fix each time we rebase but it's not mandatory to do it now.

undo-bkenward replied:
undo.go has a sufficiently complex history to make that a bit fiddly.

mark-undoio reviewed (approved):
I have some thoughts.  I suggest we just agree to address them later rather than delaying the update.

undo-bkenward committed:
Handle new session fields from recent UDB

UDB now saves some additional paths to the session file. We don't do
anything with these paths, but we need to be aware of them.
